File: code/rust/ch10/src/state/lua_table.rs
use super::lua_value::LuaValue;
use std::cell::RefCell;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};

// https://doc.rust-lang.org/std/cell/index.html#introducing-mutability-inside-of-something-immutable
#[derive(Clone)]
pub struct LuaTable {
    arr: RefCell<Vec<LuaValue>>,
    map: RefCell<HashMap<LuaValue, LuaValue>>,
    rdm: usize, // hash code
}

impl Hash for LuaTable {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.rdm.hash(state);
    }
}

impl LuaTable {
    pub fn new(narr: usize, nrec: usize) -> LuaTable {
        LuaTable {
            arr: RefCell::new(Vec::with_capacity(narr)),
            map: RefCell::new(HashMap::with_capacity(nrec)),
            rdm: super::math::random(),
        }
    }

    pub fn len(&self) -> usize {
        self.arr.borrow().len()
    }

    pub fn get(&self, key: &LuaValue) -> LuaValue {
        let arr = self.arr.borrow();
        if let Some(idx) = to_index(key) {
            if idx <= arr.len() {
                return arr[idx - 1].clone(); // TODO
            }
        }

        let map = self.map.borrow();
        if let Some(val) = map.get(key) {
            val.clone() // TODO
        } else {
            LuaValue::Nil
        }
    }

    pub fn put(&self, key: LuaValue, val: LuaValue) {
        if key.is_nil() {
            panic!("table index is nil!");
        }
        if let LuaValue::Number(n) = key {
            if n.is_nan() {
                panic!("table index is NaN!");
            }
        }

        let mut arr = self.arr.borrow_mut();
        let mut map = self.map.borrow_mut();
        if let Some(idx) = to_index(&key) {
            let arr_len = arr.len();
            if idx <= arr_len {
                let val_is_nil = val.is_nil();
                arr[idx - 1] = val;
                if idx == arr_len && val_is_nil {
                    shrink_array(&mut arr);
                }
                return;
            }
            if idx == arr_len + 1 {
                map.remove(&key);
                if !val.is_nil() {
                    arr.push(val);
                    expand_array(&mut arr, &mut map);
                }
                return;
            }
        }

        if !val.is_nil() {
            map.insert(key, val);
        } else {
            map.remove(&key);
        }
    }
}

fn to_index(key: &LuaValue) -> Option<usize> {
    if let LuaValue::Integer(i) = key {
        if *i >= 1 {
            return Some(*i as usize);
        }
    } else if let LuaValue::Number(n) = key {
        if let Some(i) = super::math::float_to_integer(*n) {
            if i >= 1 {
                return Some(i as usize);
            }
        }
    }
    None
}

fn shrink_array(arr: &mut Vec<LuaValue>) {
    while !arr.is_empty() {
        if arr.last().unwrap().is_nil() {
            arr.pop();
        } else {
            break;
        }
    }
}

fn expand_array(arr: &mut Vec<LuaValue>, map: &mut HashMap<LuaValue, LuaValue>) {
    let mut idx = arr.len() + 1;
    loop {
        let key = LuaValue::Integer(idx as i64);
        if map.contains_key(&key) {
            let val = map.remove(&key).unwrap();
            arr.push(val);
            idx += 1;
        } else {
            break;
        }
    }
}

File: code/python/ch10/src/lua_table.py
from lua_value import LuaValue


class LuaTable:
    def __init__(self, narr, nrec):
        self.arr = None
        self.map = None

        if narr > 0:
            self.arr = []
        if nrec > 0:
            self.map = {}

    def __len__(self):
        return len(self.arr) if self.arr is not None else 0

    def get(self, key):
        key = LuaValue.float2integer(key)

        if self.arr and isinstance(key, int):
            if 1 <= key <= len(self.arr):
                return self.arr[key-1]

        return self.map[key] if (self.map and key in self.map) else None

    def map_to_arr(self):
        if self.map:
            i = len(self.arr)
            while i in self.map:
                v = self.map[i]
                self.map.pop(i)
                self.arr.append(v)
                i += 1

    def put(self, key, val):
        assert(key is not None)
        assert(key is not float('NAN'))

        key = LuaValue.float2integer(key)
        if isinstance(key, int) and key >= 1:
            if not self.arr:
                self.arr = []
            arr_len = len(self.arr)
            if key <= arr_len:
                self.arr[key-1] = val
                if key == arr_len and val is None:
                    self.arr.pop(key)
                return
            if key == arr_len + 1:
                if self.map:
                    self.map.pop(key)
                if val is not None:
                    self.arr.append(val)
                    self.map_to_arr()
                return

        if val is not None:
            if not self.map:
                self.map = {key: val}
            else:
                self.map[key] = val
        else:
            self.map.pop(key)

    def dump(self):
        if self.arr:
            print(self.arr, end=' ')
        if self.map:
            print(self.map)
